utils/s3/client.cc

```diff
@@ -847,6 +847,82 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// a sink with known object size
+//
+// unlike upload_sink and upload_jumbo_sink, upload_ranged_sink sends the
+// received buffers right away without accumulating them first. caller is
+// supposed to put() with the payload of part_size until the specified
+// total_size is reached.
+class client::upload_ranged_sink final : public upload_sink_base {
+    // TODO: resume support
+    const size_t _total_size;
+    const size_t _part_size;
+    const unsigned _total_num_parts;
+
+    size_t next_put_size() const {
+        assert(_part_etags.size() <= _total_num_parts);
+        if (_part_etags.size() == _total_num_parts) {
+            return 0;
+        }
+        size_t remainder = _total_size - _part_etags.size() * _part_size;
+        return std::min(remainder, _part_size);
+    }
+
+public:
+    upload_ranged_sink(shared_ptr<client> cln,
+                       sstring object_name,
+                       size_t total_size,
+                       size_t part_size)
+        : upload_sink_base(std::move(cln), std::move(object_name), std::nullopt)
+        , _total_size{total_size}
+        , _part_size{part_size}
+        , _total_num_parts{(total_size + part_size - 1) / part_size} {
+        assert(_part_size >= aws_minimum_part_size);
+        assert(_total_num_parts <= aws_maximum_parts_in_piece);
+    }
+
+    future<> put(temporary_buffer<char> buf) final {
+        assert(_part_etags.size() < _total_num_parts);
+        assert(next_put_size() == buf.size());
+        memory_data_sink_buffers sink_buffers;
+        sink_buffers.put(std::move(buf));
+        co_await upload_part(std::move(sink_buffers));
+    }
+
+    future<> flush() final {
+        // we don't keep a partial of the multipart part in memory, so no
+        // need to flush the bits as new part. but neither do we could
+        // wait until all background flushes complete
+        return make_ready_future();
+    }
+
+    future<> close() final {
+        size_t remainder = _part_etags.size() * next_put_size();
+        if (remainder != 0) {
+            // before resume support is implemented, it is considered an
+            // unexpected usage
+            s3l.warn("premature close(): {}/{} to go", remainder, _total_size);
+            co_await upload_sink_base::close();
+        } else {
+            co_await finalize_upload();
+        }
+    }
+
+    size_t buffer_size() const noexcept final {
+        // we don't buffer at all
```

Comment:
It's not for sink buffer size, this method is called by output_stream() that's created without specifying buffer size

Reply:
yeah, but the comment still serve its purpose. i think i didn't misinterpret how this function is used.

Reply:
it's not sink's buffer, but it's for the output_stream's zero copy buffer. anyway, i dropped this override along with this controversial comment.